<a href="https://cognitiveclass.ai"><img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork/images/IDSN-logo.png" width="400"> </a>

# Regression Models with Keras

Estimated time needed: **45** mins


## Introduction


As we discussed in the videos, despite the popularity of more powerful libraries such as PyToch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the Keras library. 

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code. You will appreciate Keras even more, once you learn how to build deep models using PyTorch and TensorFlow in the other courses.

So, in this lab, you will learn how to use the Keras library to build a regression model.


## Objectives for this Notebook    
* How to use the Keras library to build a regression model
* Download and clean the data set
* Build a neural network
* Train and test the network     



## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#item31">Download and Clean the Data Set</a>  
2. <a href="#item32">Import Keras Packages</a>
3. <a href="#item33">Build a Neural Network</a>
4. <a href="#item34">Train and Test the Network</a>  

</font>
</div>


<a id="item31"></a>


## Download and Clean the  Data Set


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [1]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. 
# If you run this notebook on a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.

!pip install numpy==2.0.2
!pip install pandas==2.2.2
!pip install tensorflow_cpu==2.18.0


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.9/60.9 kB 7.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.5/19.5 MB 105.6 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.0/13.0 MB 110.2 MB/s eta 0:00:0000:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.6/346.6 kB 38.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.8/47.8 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.2/230.2 MB 2.4 MB/s eta 0:00:00:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.7/133.7 kB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.9/5.9 MB 100.8 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 65.1 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 64.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.5/24.5 MB 23.8 MB/s eta 0:00:00:00:01

##### To use Keras, you will also need to install a backend framework – such as TensorFlow.

If you install TensorFlow 2.16 or above, it will install Keras by default.

We are using the CPU version of tensorflow since we are dealing with smaller datasets. 
You may install the GPU version of tensorflow on your machine to accelarate the processing of larger datasets


In [2]:
import pandas as pd
import numpy as np
import keras

import warnings
warnings.simplefilter('ignore', FutureWarning)

2025-01-08 15:06:33.319151: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-01-08 15:06:33.366573: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

* Cement
* Blast furnace slag
* Fly ash
* Water
* Superplasticizer
* Coarse aggregate
* Fine aggregate


Let's download the data and read it into a <em>pandas</em> dataframe.


In [3]:
filepath='https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv'
concrete_data = pd.read_csv(filepath)

concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 


#### Let's check how many data points we have


In [4]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [5]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [6]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [7]:
concrete_data_columns = concrete_data.columns

In [8]:
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


Let's do a quick sanity check of the predictors and the target dataframes.


In [9]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [10]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.


In [11]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [12]:
n_cols = predictors_norm.shape[1] # number of predictors

<a id="item1"></a>


<a id='item32'></a>


##  Import Keras Packages

##### Let's import the rest of the packages from the Keras library that we will need to build our regression model.


In [13]:
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Input

<a id='item33'></a>


## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.


In [15]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Input(shape=(n_cols,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has two hidden layers, each of 50 hidden units.


<a id="item4"></a>


## Train and Test the Network


Let's call the function now to create our model.


In [17]:
# build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.


In [18]:
# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Epoch 1/100
23/23 - 1s - 49ms/step - loss: 1628.8318 - val_loss: 1121.7173
Epoch 2/100
23/23 - 0s - 8ms/step - loss: 1466.8854 - val_loss: 980.4820
Epoch 3/100
23/23 - 0s - 8ms/step - loss: 1232.2520 - val_loss: 782.2939
Epoch 4/100
23/23 - 0s - 8ms/step - loss: 918.6968 - val_loss: 545.6415
Epoch 5/100
23/23 - 0s - 7ms/step - loss: 579.2217 - val_loss: 351.1949
Epoch 6/100
23/23 - 0s - 7ms/step - loss: 348.9029 - val_loss: 246.9337
Epoch 7/100
23/23 - 0s - 8ms/step - loss: 263.6465 - val_loss: 204.9550
Epoch 8/100
23/23 - 0s - 7ms/step - loss: 235.0561 - val_loss: 189.0624
Epoch 9/100
23/23 - 0s - 8ms/step - loss: 218.0444 - val_loss: 179.1597
Epoch 10/100
23/23 - 0s - 6ms/step - loss: 207.5049 - val_loss: 173.6058
Epoch 11/100
23/23 - 0s - 7ms/step - loss: 197.4213 - val_loss: 173.2428
Epoch 12/100
23/23 - 0s - 6ms/step - loss: 190.2874 - val_loss: 169.6810
Epoch 13/100
23/23 - 0s - 7ms/step - loss: 184.3099 - val_loss: 166.7415
Epoch 14/100
23/23 - 0s - 7ms/step - loss: 178.6294 - v

<strong>You can refer to this [link](https://keras.io/models/sequential/) to learn about other functions that you can use for prediction or evaluation.</strong>


Feel free to vary the following and note what impact each change has on the model's performance:

1. Increase or decreate number of neurons in hidden layers
2. Add more hidden layers
3. Increase number of epochs


In [19]:
# Example: Change number of neurons in hidden layers
model.add(Dense(100, activation='relu'))  # Increase to 100 neurons


In [20]:
# Example: Add another hidden layer
model.add(Dense(30, activation='relu'))  # Add a new hidden layer with 30 neurons


In [21]:
from tensorflow.keras.layers import Dropout

# Example: Add dropout layer
model.add(Dropout(0.3))  # Deactivates 30% of neurons during training


In [22]:
from tensorflow.keras.callbacks import EarlyStopping

# Define early stopping
early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)

# Train with early stopping
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2, callbacks=[early_stopping])


Epoch 1/100
23/23 - 0s - 9ms/step - loss: 34.2158 - val_loss: 159.9321
Epoch 2/100
23/23 - 0s - 7ms/step - loss: 33.6960 - val_loss: 168.4626
Epoch 3/100
23/23 - 0s - 6ms/step - loss: 33.5934 - val_loss: 157.7046
Epoch 4/100
23/23 - 0s - 7ms/step - loss: 32.9779 - val_loss: 163.2965
Epoch 5/100
23/23 - 0s - 6ms/step - loss: 33.4012 - val_loss: 163.0168
Epoch 6/100
23/23 - 0s - 7ms/step - loss: 32.5414 - val_loss: 163.1573
Epoch 7/100
23/23 - 0s - 7ms/step - loss: 31.7213 - val_loss: 169.6808
Epoch 8/100
23/23 - 0s - 6ms/step - loss: 31.8583 - val_loss: 166.4399
Epoch 9/100
23/23 - 0s - 7ms/step - loss: 30.7459 - val_loss: 162.9697
Epoch 10/100
23/23 - 0s - 7ms/step - loss: 30.5546 - val_loss: 159.2968
Epoch 11/100
23/23 - 0s - 6ms/step - loss: 30.4466 - val_loss: 176.2298
Epoch 12/100
23/23 - 0s - 6ms/step - loss: 30.0935 - val_loss: 165.5340
Epoch 13/100
23/23 - 0s - 7ms/step - loss: 29.8193 - val_loss: 172.7950


In [23]:
# Train for fewer epochs (e.g., 50)
model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)


Epoch 1/50
23/23 - 0s - 8ms/step - loss: 33.5791 - val_loss: 155.1771
Epoch 2/50
23/23 - 0s - 6ms/step - loss: 32.7287 - val_loss: 163.8856
Epoch 3/50
23/23 - 0s - 7ms/step - loss: 32.3963 - val_loss: 165.6782
Epoch 4/50
23/23 - 0s - 13ms/step - loss: 31.8057 - val_loss: 162.2749
Epoch 5/50
23/23 - 0s - 7ms/step - loss: 32.0175 - val_loss: 172.3020
Epoch 6/50
23/23 - 0s - 6ms/step - loss: 31.3730 - val_loss: 170.9568
Epoch 7/50
23/23 - 0s - 7ms/step - loss: 30.7469 - val_loss: 166.0178
Epoch 8/50
23/23 - 0s - 7ms/step - loss: 30.3296 - val_loss: 168.8950
Epoch 9/50
23/23 - 0s - 7ms/step - loss: 30.2536 - val_loss: 164.9199
Epoch 10/50
23/23 - 0s - 6ms/step - loss: 30.0231 - val_loss: 174.2541
Epoch 11/50
23/23 - 0s - 8ms/step - loss: 30.0887 - val_loss: 180.0644
Epoch 12/50
23/23 - 0s - 6ms/step - loss: 30.0147 - val_loss: 172.9956
Epoch 13/50
23/23 - 0s - 7ms/step - loss: 29.0711 - val_loss: 166.6403
Epoch 14/50
23/23 - 0s - 6ms/step - loss: 28.7097 - val_loss: 162.6844
Epoch 15/50
23

In [26]:
!pip install scikit-learn


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.0/62.0 kB 7.2 kB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.5/13.5 MB 52.1 MB/s eta 0:00:00:00:010:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 301.8/301.8 kB 13.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.6/40.6 MB 12.1 MB/s eta 0:00:00:00:0100:01


In [27]:
from sklearn.model_selection import train_test_split

# Split predictors and target into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.2, random_state=42)


In [29]:
def regression_model():
    # Create the model
    model = Sequential()
    model.add(Dense(50, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(1))  # Output layer for regression
    
    # Compile the model with metrics
    model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])
    return model


In [31]:
def regression_model():
    # Create the model
    model = Sequential()
    model.add(Dense(50, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(1))  # Output layer for regression

    # Compile the model with metrics
    model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])
    return model


In [32]:
# Recreate the model
model = regression_model()

# Train the model
history = model.fit(X_train, y_train, validation_split=0.2, epochs=100, verbose=2)


Epoch 1/100


/opt/conda/lib/python3.11/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


21/21 - 1s - 56ms/step - loss: 1529.6016 - mae: 35.1339 - val_loss: 1453.6139 - val_mae: 34.9137
Epoch 2/100
21/21 - 0s - 8ms/step - loss: 1413.1160 - mae: 33.4648 - val_loss: 1318.3593 - val_mae: 32.9797
Epoch 3/100
21/21 - 0s - 8ms/step - loss: 1237.5818 - mae: 30.8434 - val_loss: 1109.4183 - val_mae: 29.7669
Epoch 4/100
21/21 - 0s - 8ms/step - loss: 980.9515 - mae: 26.6439 - val_loss: 822.5424 - val_mae: 24.9384
Epoch 5/100
21/21 - 0s - 7ms/step - loss: 672.5818 - mae: 21.1260 - val_loss: 511.6415 - val_mae: 18.7938
Epoch 6/100
21/21 - 0s - 7ms/step - loss: 408.6931 - mae: 15.8428 - val_loss: 291.7475 - val_mae: 13.5940
Epoch 7/100
21/21 - 0s - 7ms/step - loss: 269.8975 - mae: 12.9614 - val_loss: 208.7781 - val_mae: 11.6557
Epoch 8/100
21/21 - 0s - 7ms/step - loss: 237.6214 - mae: 12.2705 - val_loss: 182.9186 - val_mae: 11.0680
Epoch 9/100
21/21 - 0s - 8ms/step - loss: 220.8341 - mae: 11.8538 - val_loss: 174.1792 - val_mae: 10.9074
Epoch 10/100
21/21 - 0s - 7ms/step - loss: 211.6877

In [33]:
# Evaluate the model
results = model.evaluate(X_test, y_test, verbose=0)

# Check results
if isinstance(results, (list, tuple)):
    loss, mae = results
    print(f"Test Loss: {loss}")
    print(f"Test Mean Absolute Error: {mae}")
else:
    print(f"Test Loss: {results}")  # If only loss is returned


Test Loss: 68.01287841796875
Test Mean Absolute Error: 6.56852388381958


In [34]:
# Predict on test data
predictions = model.predict(X_test)
print(predictions[:5])  # Display the first 5 predictions


7/7 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
[[48.865204]
 [49.19633 ]
 [67.3029  ]
 [48.099262]
 [16.804104]]


In [36]:
# Combine actual and predicted values for comparison
import pandas as pd

comparison = pd.DataFrame({'Actual': y_test[:5].values, 'Predicted': predictions[:5].flatten()})
print(comparison)


   Actual  Predicted
0   52.91  48.865204
1   55.90  49.196331
2   74.50  67.302902
3   35.30  48.099262
4   10.54  16.804104


In [37]:
!pip install matplotlib


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 165.1/165.1 kB 22.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.6/8.6 MB 132.6 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 326.2/326.2 kB 47.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 120.8 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 89.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 114.3 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 107.7/107.7 kB 15.2 MB/s eta 0:00:00


In [38]:
# Example: Plot training and validation loss
plt.plot(history.history['loss'], label='Training Loss')
plt.plot(history.history['val_loss'], label='Validation Loss')
plt.xlabel('Epochs')
plt.ylabel('Loss')
plt.legend()
plt.title('Model Loss Over Epochs')
plt.show()


NameError: name 'plt' is not defined

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/). I hope you found this lab interesting and educational. Feel free to contact me if you have any questions!


<!--
## Change Log

|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2024-11-20  | 3.0  | Aman  |  Updated the library versions to current |
| 2020-09-21  | 2.0  | Srishti  |  Migrated Lab to Markdown and added to course repo in GitLab |



<hr>

## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>


<hr>

Copyright &copy; [IBM Developer Skills Network](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://opensource.org/license/mit).
